build_runner/lib/src/asset/cache.dart

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:build/build.dart';
+import 'package:glob/glob.dart';
+
+/// An [AssetReader] that wraps another [AssetReader] and caches all results
+/// from it.
+///
+/// Assets are cached until [invalidate] is invoked.
+///
+/// Does not implement [findAssets].
+class CachingAssetReader implements AssetReader {
+  /// Cached results of [canRead].
+  final _canReadCache = <AssetId, Future<bool>>{};
+
+  /// Cached results of [readAsBytes].
+  final _bytesContentCache = <AssetId, Future<List<int>>>{};
+
+  /// Cached results of [readAsString], per [Encoding] type used.
+  ///
+  /// These are computed and stored lazily using [readAsBytes].
+  final _stringContentCache = <AssetId, Map<Encoding, Future<String>>>{};
+
+  /// The [AssetReader] to delegate all reads to.
+  final AssetReader _delegate;
+
+  CachingAssetReader(this._delegate);
+
+  @override
+  Future<bool> canRead(AssetId id) =>
```

Comment:
how do we account for files that get added and can be read later? Is this cache per-phase?

Reply:
We still wrap this in the `SinglePhaseReader`, which handles checking phase numbers etc. This just wraps the `FileBasedAssetReader` (or whatever reader you pass in to watch/build), which is always just hitting the filesystem directly.

Reply:
does that mean that new files can never appear readable? Does the SinglePhaseReader no call through for `canRead`?

Reply:
I think we should maybe only cache the `true` values and call through the delegate for everything else here.

Reply:
Ah, I didn't think that we ever called through to the underlying reader for this in the `SinglePhaseAssetReader`, but it looks like we do.

It seems like we should just be going directly to the `AssetGraph`, and checking if the `GeneratedAssetNode.wasOutput` for generated nodes. For source nodes this logic should hold fine and is actually what we want. We don't want this to return a different answer for two actions because the user added a new file during the build for instance.

Reply:
I'm a little bit nervous about having this cached like this and depending on it being wrapped in an implementation that won't call it... Can we do the safer implementation, or change it to throw not implemented?

Reply:
I agree this seems a bit tricky, but in reality it is not bad imo. This class does not in fact depend on being wrapped in a different implementation. It just does exactly what it claims to, which is cache results until you call invalidate. This is important in order to maintain the same state throughout a build.

The only special logic involved is in the `SinglePhaseAssetReader`. That class is already pretty special, and uses a lot more knowledge about the asset graph and state of the world to block reads, run builds, and return different results based on phases etc. This is just another case for it, where for generated assets it does not call the wrapped reader, but instead looks at the asset graph directly to determine if a generated file exists.

Throwing for `canRead` isn't an option because it does need to be able to work (for source assets). The other behavior of not caching for `canRead` is not actually safer, its explicitly creating a potential for inconsistent builds (if you create a new source file during a build).